xwiki-platform-core/xwiki-platform-repository/xwiki-platform-repository-server-api/src/main/java/org/xwiki/repository/internal/RepositoryManager.java

```diff
@@ -996,9 +1033,66 @@ private boolean updateExtensionVersionDependencies(XWikiDocument document, Exten
         return needSave;
     }
 
+    public Extension resolveExtensionVersion(XWikiDocument extensionDocument, String extensionVersion) throws ResolveException {
+        BaseObject extensionObject = extensionDocument.getXObject(XWikiRepositoryModel.EXTENSION_CLASSREFERENCE);
+        if(extensionObject == null){
+            return null;
+        }
+        String extensionId = getValue(extensionObject, XWikiRepositoryModel.PROP_EXTENSION_ID, (String) null);
+
+        BaseObject extensionProxyObject = extensionDocument.getXObject(XWikiRepositoryModel.EXTENSIONPROXY_CLASSREFERENCE);
+        if(extensionProxyObject == null){
+            return null;
+        }
+        String repositoryId = getValue(extensionProxyObject, XWikiRepositoryModel.PROP_PROXY_REPOSITORYID, (String) null);
+
+        if(extensionId == null || repositoryId == null){
+            return null;
+        }
+
+        ExtensionRepository repository = this.extensionRepositoryManager.getRepository(repositoryId);
```

Comment:
The current (latest version) object (ExtensionClass) nor related version and dependency objects (ExtensionVersionClass, ExtensionDependencyClass) nor proxy object (ExtensionProxyClass) don't store the info about the previous versions. In other words - to figure out what are the historic versions of given extension, it's required to ask remote repository - what I do in line 1053.

What I can optimize (and will be very useful optimization) is to check if requested version is current version. If so, all data is cached in backing xObjects of extensionDocument and no need to call remote repository.

Does it make sense, what I wrote?